### PR TITLE
Question types

### PR DESCRIPTION
**Pull request by liamwheen**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 all: questions.done
 
-%.done: %.pdf %_bb.txt
+%.done: %.pdf %_bb.txt txt2bb.py
```
Comment (oscarbenjamin) [inline Makefile]: Why have you added this?

Comment (oscarbenjamin) [reply Makefile]: If you edit the script you can use `make -B` to force recompiling everything

File: README.md [hunk 2]
```diff
@@ -35,12 +35,36 @@ incorrect: $C\sin{t}$
...
+provided questions.txt file for examples. Explanations on how each question
+type is formatted can be found
+[here](https://www.csustan.edu/sites/default/files/blackboard/FacultyHelp/Documents/UploadingQuestions.pdf).
+
+### NOTES:
+
+If the question should contain a linebreak, start each new line with `>`. 
+
+
+If you wish to ignore a linebreak in the input file preced the line with `<`.
```
Comment (oscarbenjamin) [inline README.md]: It is more common to use `\` for this kind of line continuation

File: README.md [hunk 3]
```diff
@@ -35,12 +35,36 @@ incorrect: $C\sin{t}$
...
+For question type `NUM`, there is the option to add a tolerance for the answer
+which is preceded by `tolerance: ` in the questions file.
+
+For question type `MAT`, pairs must be preceded by `match_a` and `match_b`.
 
 Usage
 -----
 
-Once you have your questions written up ou can then convert them eihter to
+Once you have your questions written up you can then convert them eihter to
```
Comment (oscarbenjamin) [inline README.md]: eihter -> either

File: README.md [hunk 4]
```diff
@@ -72,5 +96,8 @@ You can now upload questions_bb.txt to Blackboard as a test or question pool
 using the "Upload questions" button. The latex embedded with double dollar
 e.g. `$$y = x^2$$` in questions_bb.txt will not automatically render when
 uploaded as rendering latex is a feature of the web editor. This means that
-after uploading you have to edit each question and then save without chaging
-anything to triger hte latex rendering (maybe there is a better way...).
+after uploading you have to edit each question and then save without changing
+anything to trigger the latex rendering (maybe there is a better way...).
+[This help
+page](https://blackboard.secure.force.com/publickbarticleview?id=kA339000000L6QH)
+from blackboard seems to suggest there is not another way.
```
Comment (oscarbenjamin) [inline README.md:116]: There is the possibility of solving this by rendering offline if we can support the other Blackboard format though. Probably worth mentioning here.

File: q_types.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python3
+
+"""
+Class for each question type available on Blackboard. Fill-in question
+types are not included
+"""
+import re
+
+class Q_process:
```
Comment (oscarbenjamin) [inline q_types.py]: Probably `Question` is a better name for this class.

File: q_types.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    
+    questions (dict) -> assigns generic values to object
+    """
+
+    def __init__(self, question):
+        self.question = question
+        self.type = question['type']
+        self.prompt = question['prompt']
+
+class MC(Q_process):
```
Comment (oscarbenjamin) [inline q_types.py]: Can you add a docstring at the top of this module explaining what the classes are and the general idea of the code and linking to a source that explains the questions types and what the abbreviations "MC" etc mean

File: q_types.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        else:
+            return [self.type, self.prompt, self.ans]
+
+    def latex(self):
+        if 'tolerance' in self.question.keys():
+            return [('answer', self.ans), ('tolerance', r'$\pm$'+self.tol)]
+        else:
+            return [('answer', self.ans)]
+class SR(ESS):
+    """Inherets functions form essay question class as they are identical"""
```
Comment (oscarbenjamin) [inline q_types.py]: There should be 2 blank lines above this class. inherets -> inherits

The docstring is more of a code comment than documentation. Better just to use a comment

File: txt2bb.py [hunk 8]
```diff
@@ -107,22 +119,19 @@ def q2bb(questions):
...
-
+    items = handler.bb()
     # Output must be tab-delimited
     # Need double-$ in Blackboard
     return "\t".join(items).replace("$", "$$")
 
 
 LATEX_START = r"""
 \documentclass{article}
+\usepackage{amsmath}
```
Comment (oscarbenjamin) [inline txt2bb.py:366]: We can't use amsmath in blackboard though can we?

File: q_types.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python3
```
Comment (oscarbenjamin) [inline q_types.py]: I don't think we need a separate file for this code. It's probably good to keep all the actual code in one file for now

File: README.md [hunk 10]
```diff
@@ -35,12 +35,36 @@ incorrect: $C\sin{t}$
...
+provided questions.txt file for examples. Explanations on how each question
+type is formatted can be found
+[here](https://www.csustan.edu/sites/default/files/blackboard/FacultyHelp/Documents/UploadingQuestions.pdf).
+
+### NOTES:
+
+If the question should contain a linebreak, start each new line with `>`. 
+
+
+If you wish to ignore a linebreak in the input file preced the line with `\`.
```
Comment (oscarbenjamin) [inline README.md]: preced -> precede

File: README.md [hunk 11]
```diff
@@ -35,12 +35,36 @@ incorrect: $C\sin{t}$
...
+
+For question type `JUMBLED_SENTENCE`, the choices are listed as answers,
```
Comment (oscarbenjamin) [inline README.md:64]: Start by saying that examples of each type can be seen in questions.txt

File: README.md (hunk 10 above)
Comment (oscarbenjamin) [reply README.md]: It's probably best to give examples of using `>` and `\` to show how it looks

File: questions.txt [hunk 12]
```diff
@@ -19,3 +19,78 @@ correct: $C\mathrm{e}^t$
...
+
+------------------ Question
+
+type: FIL
+prompt: Upload a file please
+
+------------------ Question
+
+type: NUM
+prompt: Out of 10 how good is this question
```
Comment (oscarbenjamin) [inline questions.txt]: ?

File: questions.txt [hunk 13]
```diff
@@ -19,3 +19,78 @@ correct: $C\mathrm{e}^t$
...
+------------------ Question
+
+type: SR
+prompt: In a few words, how is this going
+example: very well
+
+------------------ Question
+
+type: OP
+prompt: How good is this going so far?
```
Comment (oscarbenjamin) [inline questions.txt]: Is there a blackboard doc giving a good explanation of what these question types mean or are for? Can we link to that?

File: txt2bb.py [hunk 14]
```diff
@@ -28,42 +28,288 @@
...
+IN_TYPES = ('correct', 'incorrect', 'answer', 'match_a', 'match_b', 'example',
+        'tolerance', 'variable', 'q_word', 'q_phrase')        
+HANDLERS = dict(zip(Q_TYPES, q_handlers))
 
 def main(mode, filename):
     """ $ ./txt2bb.py (--latex|--b) FILE """
 
     with open(filename) as infile:
-        questions = txt2py(infile)
+        questions = txt2py(mode,infile)
```
Comment (oscarbenjamin) [inline txt2bb.py]: Why do we need to know the mode when reading the file?

Comment (oscarbenjamin) [reply txt2bb.py]: Also space after comma (pep8)

Comment (oscarbenjamin) [reply txt2bb.py]: Output mode should be handled at output not input

File: questions.txt (hunk 13 above)
Comment (liamwheen) [reply questions.txt]: yes its at the top of the file:

> Each class corresponds to one of Blackboard's question types, for example, MC
is multiple choice. Full documentation for each question type and its format
can be found here:

https://www.csustan.edu/sites/default/files/blackboard/FacultyHelp/Documents/UploadingQuestions.pdf

>

File: questions.txt (hunk 12 above)
Comment (liamwheen) [reply questions.txt]: If you're asking why its not maths, as I said, theres still some more questions to write but if there isn't a hurry I'll do it tomorrow.

File: questions.txt (hunk 13 above)
Comment (oscarbenjamin) [reply questions.txt]: That doesn't explain what they *mean*. It just says how to format them in the csv file

File: questions.txt (hunk 12 above)
Comment (oscarbenjamin) [reply questions.txt]: There's no hurry. I was pointing out the missing question mark

Comment (liamwheen) [reply questions.txt]: gotcha solved the issue of the missing question mark

File: questions.txt (hunk 13 above)
Comment (liamwheen) [reply questions.txt]: I think its more clear now. Theres a link for what the questions are, and the link for what each question's code is e.g. MC

File: questions.txt [hunk 15]
```diff
@@ -19,3 +19,94 @@ correct: $C\mathrm{e}^t$
...
+>Order the following probability measures with the largest at the top.
+answer: The mode of $X$
+answer: The median of $X$
+answer: The mean of $X$
+
+------------------ Question
+
+type: MAT
+prompt: For the complex number, $z = 2+3j$, match the following expressions to
+\their corresponding value.
```
Comment (oscarbenjamin) [inline questions.txt]: Using `\` like this is confusing. Normally `\` is used as a line continuation character at the end of the line e.g.:
```
this line continues \
below.
```
Also can we make actual line breaks using <br> or something?

**Pull request merged**